File: scala/com/kotori316/fluidtank/render/ItemModelTank.scala
package com.kotori316.fluidtank.render

import java.util.Collections
import javax.vecmath.{Matrix4f, Vector3f}

import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.block.model.ItemCameraTransforms.TransformType
import net.minecraft.client.renderer.block.model.{IBakedModel, ItemOverrideList}
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.model.TRSRTransformation
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import org.apache.commons.lang3.tuple.Pair

@SideOnly(Side.CLIENT)
class ItemModelTank extends IBakedModel {

    val transMap = Map(
        TransformType.GUI -> new TRSRTransformation(null, fromDegrees(30, 135, 0), vec(0.625), null),
        TransformType.FIXED -> new TRSRTransformation(null, fromDegrees(0, 90, 0), vec(0.5), null),
        TransformType.FIRST_PERSON_LEFT_HAND -> new TRSRTransformation(null, fromDegrees(0, 45, 0), vec(0.4), null),
        TransformType.FIRST_PERSON_RIGHT_HAND -> new TRSRTransformation(null, fromDegrees(0, 225, 0), vec(0.4), null),
        TransformType.THIRD_PERSON_LEFT_HAND -> new TRSRTransformation(new Vector3f(0, 2.5f / 16, 0), fromDegrees(75, 45, 0), new Vector3f(-.375f, .375f, .375f), null),
        TransformType.THIRD_PERSON_RIGHT_HAND -> new TRSRTransformation(new Vector3f(0, 2.5f / 16, 0), fromDegrees(75, 45, 0), vec(0.375), null),
        TransformType.GROUND -> new TRSRTransformation(new Vector3f(0, 3f / 16, 0), null, vec(0.25), null)).map {
        case (t, trans) => (t, trans)
    }

    private def fromDegrees(x: Int, y: Int, z: Int) = TRSRTransformation.quatFromXYZDegrees(new Vector3f(x.toFloat, y.toFloat, z.toFloat))

    private def vec(t: Double) = new Vector3f(t.toFloat, t.toFloat, t.toFloat)

    override def isGui3d = false

    override def getOverrides = ItemOverrideList.NONE

    override def isAmbientOcclusion = false

    override def getQuads(state: IBlockState, side: EnumFacing, rand: Long) = Collections.emptyList()

    override def isBuiltInRenderer = true

    override def getParticleTexture = null

    override def handlePerspective(cameraTransformType: TransformType): Pair[_ <: IBakedModel, Matrix4f] = {
        //        PerspectiveMapWrapper.handlePerspective(this, transMap, cameraTransformType)
        val tr = transMap.getOrElse(cameraTransformType, TRSRTransformation.identity())
        Pair.of(this, if (!tr.isIdentity) tr.getMatrix else null)
    }
}

File: src/main/scala/com/kotori316/fluidtank/render/ItemModelTank.java
package com.kotori316.fluidtank.render;

import java.util.Collections;
import java.util.List;
import java.util.Random;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import net.minecraft.block.BlockState;
import net.minecraft.client.renderer.Vector3f;
import net.minecraft.client.renderer.model.BakedQuad;
import net.minecraft.client.renderer.model.ItemOverrideList;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.util.Direction;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.client.model.data.IDynamicBakedModel;
import net.minecraftforge.client.model.data.IModelData;

@OnlyIn(Dist.CLIENT)
public class ItemModelTank implements IDynamicBakedModel {
    @Nonnull
    @Override
    public List<BakedQuad> getQuads(@Nullable BlockState state, @Nullable Direction side, @Nonnull Random rand, @Nonnull IModelData extraData) {
        return Collections.emptyList();
    }

    @Override
    public boolean isAmbientOcclusion() {
        return false;
    }

    @Override
    public boolean isGui3d() {
        return false;
    }

    @Override
    public boolean func_230044_c_() {
        return true;
    }

    @Override
    public boolean isBuiltInRenderer() {
        return true;
    }

    @Override
    @Deprecated
    public TextureAtlasSprite getParticleTexture() {
        return null;
    }

    @Override
    public ItemOverrideList getOverrides() {
        return ItemOverrideList.EMPTY;
    }

    @Override
    @SuppressWarnings("deprecation")
    public net.minecraft.client.renderer.model.ItemCameraTransforms getItemCameraTransforms() {
        return Constant.transform;
    }

    @SuppressWarnings("deprecation")
    private static class Constant {
        private static final net.minecraft.client.renderer.model.ItemCameraTransforms transform = new net.minecraft.client.renderer.model.ItemCameraTransforms(
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(75, 45, 0), new Vector3f(0, 2.5f / 16, 0), vec(0.375)), //third person_leftIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(75, 45, 0), new Vector3f(0, 2.5f / 16, 0), vec(0.375)), //third person_rightIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(0, 45, 0), net.minecraft.client.renderer.model.ItemTransformVec3f.Deserializer.TRANSLATION_DEFAULT, vec(0.4)), //first person_leftIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(0, 225, 0), net.minecraft.client.renderer.model.ItemTransformVec3f.Deserializer.TRANSLATION_DEFAULT, vec(0.4)), //first person_rightIn,
            net.minecraft.client.renderer.model.ItemTransformVec3f.DEFAULT, //headIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(30, 135, 0), net.minecraft.client.renderer.model.ItemTransformVec3f.Deserializer.TRANSLATION_DEFAULT, vec(0.625)), //guiIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(net.minecraft.client.renderer.model.ItemTransformVec3f.Deserializer.ROTATION_DEFAULT, new Vector3f(0, 3f / 16, 0), vec(0.25)), //groundIn,
            new net.minecraft.client.renderer.model.ItemTransformVec3f(new Vector3f(0, 90, 0), net.minecraft.client.renderer.model.ItemTransformVec3f.Deserializer.TRANSLATION_DEFAULT, vec(0.5)) //fixedIn
        );

        private static Vector3f vec(double d) {
            return new Vector3f((float) d, (float) d, (float) d);
        }
    }
}
